### PR TITLE
chore(coderd): improve tests for tar<->zip conversion

### PR DESCRIPTION
**Pull request by johnstcn**
STACK:
https://github.com/coder/coder/pull/12477 <-- you are here
https://github.com/coder/coder/pull/12476
https://github.com/coder/coder/pull/12475
https://github.com/coder/coder/pull/12479

- Adds some artisanally handcrafted zip and tar archives under `coderd/testdata`
- Adds test helpers to assert content of above archives
- Updates existing tests to reference above archives

NOTE: these tests will fail until https://github.com/coder/coder/pull/12476 is merged in

### CODE REVIEW

File: coderd/fileszip_test.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			_ = rc.Close()
+			require.NoError(t, err)
+			require.Equal(t, "world", string(bs))
+		default:
+			require.Failf(t, "unexpected file in zip", f.Name)
+		}
+	}
+}
+
+func must[T any](t T, err error) T {
```
Comment (mafredri) [inline coderd/fileszip_test.go:212]: I'd suggest taking T too and using `t.Fatal` instead of panic.

File: coderd/files_test.go [hunk 2]
```diff
@@ -101,14 +106,7 @@ func TestDownload(t *testing.T) {
...
 		// given
-		tarball, err := echo.Tar(&echo.Responses{
-			Parse:          echo.ParseComplete,
-			ProvisionApply: echo.ApplyComplete,
-		})
-		require.NoError(t, err)
-
-		tarReader := tar.NewReader(bytes.NewReader(tarball))
-		zipContent, err := coderd.CreateZipFromTar(tarReader)
+		zipContent, err := os.ReadFile(filepath.Join(".", "testdata", "test.zip"))
```
Comment (mafredri) [inline coderd/files_test.go:109]: Nit: Simplify, it's already relative without the `.`.

```suggestion
		zipContent, err := os.ReadFile(filepath.Join("testdata", "test.zip"))
```

File: coderd/files_test.go [hunk 3]
```diff
@@ -83,16 +84,20 @@ func TestDownload(t *testing.T) {
 		// given
 		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
 		defer cancel()
+		tarball, err := os.ReadFile(filepath.Join(".", "testdata", "test.tar"))
```
Comment (mafredri) [inline coderd/files_test.go:87]:
```suggestion
		tarball, err := os.ReadFile(filepath.Join("testdata", "test.tar"))
```

File: coderd/files_test.go [hunk 4]
```diff
@@ -122,7 +120,11 @@ func TestDownload(t *testing.T) {
 
 		// then
 		require.Equal(t, codersdk.ContentTypeTar, contentType)
-		require.Equal(t, tarball, data)
+
+		// Note: creating a zip from a tar will result in some loss of information
+		// as zip files do not store UNIX user:group data.
+		// require.Equal(t, tarball, data)
```
Comment (mafredri) [inline coderd/files_test.go:126]: If we never can enable this piece of code, maybe we just remove it?
```suggestion
```

File: coderd/files_test.go [hunk 5]
```diff
@@ -131,10 +133,7 @@ func TestDownload(t *testing.T) {
 		_ = coderdtest.CreateFirstUser(t, client)
 
 		// given
-		tarball, err := echo.Tar(&echo.Responses{
-			Parse:          echo.ParseComplete,
-			ProvisionApply: echo.ApplyComplete,
-		})
+		tarball, err := os.ReadFile(filepath.Join(".", "testdata", "test.tar"))
```
Comment (mafredri) [inline coderd/files_test.go:136]:
```suggestion
		tarball, err := os.ReadFile(filepath.Join("testdata", "test.tar"))
```

File: coderd/fileszip_test.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+func TestCreateTarFromZip(t *testing.T) {
+	t.Parallel()
+	if runtime.GOOS != "linux" {
+		t.Skip("skipping this test on non-Linux platform")
+	}
+
+	// Read a zip file we prepared earlier
+	ctx := testutil.Context(t, testutil.WaitShort)
+	zipBytes, err := os.ReadFile(filepath.Join(".", "testdata", "test.zip"))
```
Comment (mafredri) [inline coderd/fileszip_test.go:37]:
```suggestion
	zipBytes, err := os.ReadFile(filepath.Join("testdata", "test.zip"))
```

File: coderd/fileszip_test.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+)
+
+// refTime is the Go reference time. The contents of the sample tar and zip files
+// in testdata/ all have their modtimes set to the below in some timezone.
+var refTime = time.Date(2006, 1, 2, 3, 4, 5, 0, must(time.LoadLocation("MST")))
+
+func TestCreateTarFromZip(t *testing.T) {
+	t.Parallel()
+	if runtime.GOOS != "linux" {
+		t.Skip("skipping this test on non-Linux platform")
```
Comment (mafredri) [inline coderd/fileszip_test.go:32]: Windows is no bueno?

File: coderd/fileszip_test.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+	"github.com/stretchr/testify/require"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd"
+	"github.com/coder/coder/v2/testutil"
+)
+
+// refTime is the Go reference time. The contents of the sample tar and zip files
+// in testdata/ all have their modtimes set to the below in some timezone.
+var refTime = time.Date(2006, 1, 2, 3, 4, 5, 0, must(time.LoadLocation("MST")))
```
Comment (mafredri) [inline coderd/fileszip_test.go:27]: Nit: Ambiguous variable name in huge package, perhaps `archiveRefTime`?

File: coderd/fileszip_test.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+	assertExtractedFiles(t, tempDir, true)
+}
+
+func TestCreateZipFromTar(t *testing.T) {
+	t.Parallel()
+	if runtime.GOOS != "linux" {
+		t.Skip("skipping this test on non-Linux platform")
+	}
+
+	tarBytes, err := os.ReadFile(filepath.Join(".", "testdata", "test.tar"))
```
Comment (mafredri) [inline coderd/fileszip_test.go:66]:
```suggestion
	tarBytes, err := os.ReadFile(filepath.Join("testdata", "test.tar"))
```

File: coderd/fileszip_test.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+				assert.Equal(t, fs.ModePerm&0o644, stat.Mode().Perm(), "expected mode 0644 on file")
+			}
+			bs, err := os.ReadFile(path)
+			assert.NoError(t, err, "failed to read file %q", path)
+			assert.Equal(t, "world", string(bs), "unexpected content in file %q", path)
+		default:
+			assert.Fail(t, "unexpected path", relPath)
+		}
+
+		t.Logf(relPath)
```
Comment (mafredri) [inline coderd/fileszip_test.go:134]: Still needed?

**Review by mafredri — APPROVED** [review]
Just minor nits, nice thorough tests!

File: coderd/fileszip_test.go (hunk 7 above)
Comment (johnstcn) [reply coderd/fileszip_test.go:32]: Generally yes, but I'm not going to assume the presence of `tar` and `zip` on Windows for this test :-)

File: coderd/fileszip_test.go (hunk 1 above)
Comment (johnstcn) [reply coderd/fileszip_test.go]: moved to a func(*testing.T)

File: coderd/fileszip_test.go (hunk 8 above)
Comment (johnstcn) [reply coderd/fileszip_test.go]: updated

**Pull request merged**